File: FWFramework/Classes/Component/Calendar/FWAuthorizeCalendar.m
//
//  FWAuthorize.m
//  FWFramework
//
//  Created by wuyong on 17/3/15.
//  Copyright © 2018年 wuyong.site. All rights reserved.
//

#import "FWAuthorizeCalendar.h"
#import <EventKit/EventKit.h>
#if FWFrameworkSPM
@import FWFramework;
#else
#import "FWAuthorize.h"
#endif

@interface FWAuthorizeEventKit : NSObject <FWAuthorizeProtocol>

@property (nonatomic, readonly) EKEntityType type;

- (instancetype)initWithType:(EKEntityType)type;

@end

@implementation FWAuthorizeEventKit

+ (void)load
{
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        [FWAuthorizeManager registerAuthorize:FWAuthorizeTypeCalendars withBlock:^id<FWAuthorizeProtocol>{
            return [[FWAuthorizeEventKit alloc] initWithType:EKEntityTypeEvent];
        }];
        [FWAuthorizeManager registerAuthorize:FWAuthorizeTypeReminders withBlock:^id<FWAuthorizeProtocol>{
            return [[FWAuthorizeEventKit alloc] initWithType:EKEntityTypeReminder];
        }];
    });
}

- (instancetype)initWithType:(EKEntityType)type
{
    self = [super init];
    if (self) {
        _type = type;
    }
    return self;
}

- (FWAuthorizeStatus)authorizeStatus
{
    EKAuthorizationStatus status = [EKEventStore authorizationStatusForEntityType:self.type];
    switch (status) {
        case EKAuthorizationStatusRestricted:
            return FWAuthorizeStatusRestricted;
        case EKAuthorizationStatusDenied:
            return FWAuthorizeStatusDenied;
        case EKAuthorizationStatusAuthorized:
            return FWAuthorizeStatusAuthorized;
        case EKAuthorizationStatusNotDetermined:
        default:
            return FWAuthorizeStatusNotDetermined;
    }
}

- (void)authorize:(void (^)(FWAuthorizeStatus))completion
{
    EKEventStore *store = [[EKEventStore alloc] init];
    [store requestAccessToEntityType:self.type completion:^(BOOL granted, NSError * _Nullable error) {
        FWAuthorizeStatus status = granted ? FWAuthorizeStatusAuthorized : FWAuthorizeStatusDenied;
        if (completion) {
            dispatch_async(dispatch_get_main_queue(), ^{
                completion(status);
            });
        }
    }];
}

@end

File: FWFramework/Classes/Module/Calendar/FWAuthorizeCalendar.swift
//
//  FWAuthorizeCalendar.swift
//  FWFramework
//
//  Created by wuyong on 2019/6/27.
//  Copyright © 2019 wuyong.site. All rights reserved.
//

import EventKit
#if FWFrameworkSPM
import FWFramework
import FWFrameworkCompatible
#endif

/// 日历授权
@objcMembers public class FWAuthorizeCalendar: NSObject, FWAuthorizeProtocol, FWAutoloadProtocol {
    private var type: EKEntityType = .event
    
    public static func autoload() {
        FWAuthorizeManager.registerAuthorize(.calendars) {
            return FWAuthorizeCalendar(type: .event)
        }
        FWAuthorizeManager.registerAuthorize(.reminders) {
            return FWAuthorizeCalendar(type: .reminder)
        }
    }
    
    public init(type: EKEntityType) {
        super.init()
        self.type = type
    }
    
    public func authorizeStatus() -> FWAuthorizeStatus {
        let status = EKEventStore.authorizationStatus(for: type)
        switch status {
        case .restricted:
            return .restricted
        case .denied:
            return .denied
        case .authorized:
            return .authorized
        default:
            return .notDetermined
        }
    }
    
    public func authorize(_ completion: ((FWAuthorizeStatus) -> Void)?) {
        EKEventStore().requestAccess(to: type) { granted, error in
            let status: FWAuthorizeStatus = granted ? .authorized : .denied
            if completion != nil {
                DispatchQueue.main.async {
                    completion?(status)
                }
            }
        }
    }
}
